tests/test_backend_failure.py

```diff
@@ -0,0 +1,219 @@
+                            DownloadInterruptionHandler(number_of_failures=4),
+                            MalformedJsonHandler(number_of_failures=4)
+                        ])
+@with_secondary()
+@with_aktualizr(run_mode='full')
+@with_install_manager()
+def test_backend_failure_sanity_imagerepo_update_after_metadata_download_failure(install_mngr, director, **kwargs):
+    director.wait_for_install()
+    return director.is_install_successful() and install_mngr.are_images_installed()
+
+
+"""
+Verifies whether aktualizr is updatable after image download failure
+with follow-up successful download.
+
+Currently, it's tested against two types of image download failure:
+    - download interruption - file download is interrupted once, after that it's successful
+    - malformed image - image download is successful but it's malformed. It happens at once after that it's successful
+"""
+@with_uptane_backend(start_generic_server=True, port=8888)
+@with_images(images_to_install=[(('primary-hw-ID-001', 'primary-ecu-id'), 'primary-image.img')])
+@with_director()
+@with_imagerepo(handlers=[
+                            # TODO: test fails because aktualizr issues byte range request
+                            # that are not supported by server
+                            #DownloadInterruptionHandler(number_of_failures=1, url='/targets/primary-image.img'),
+                            MalformedImageHandler(number_of_failures=1, url='/targets/primary-image.img'),
+                        ])
+@with_secondary()
+@with_aktualizr(run_mode='full', id=('primary-hw-ID-001', 'primary-ecu-id'), output_logs=True)
+@with_install_manager()
+def test_backend_failure_sanity_imagerepo_update_after_image_download_failure(install_mngr, director, **kwargs):
+    director.wait_for_install()
+    return director.is_install_successful() and install_mngr.are_images_installed()
+
+
+"""
+    Verifies whether aktualizr is updatable after malformed image is downloaded
+    from a custom image server with follow-up successful download.
+"""
+@with_uptane_backend(start_generic_server=True, port=8888)
+@with_images(images_to_install=[(('primary-hw-ID-001', 'primary-ecu-id'), 'primary-image.img',
+                                 'http://localhost:9001/primary-image.img')])
+@with_director()
+@with_customrepo(handlers=[
+                            # TODO: This test fails because the issue with image download
+                            #  from a server that doesn't support byte range requests
+                            #DownloadInterruptionHandler(number_of_failures=1, url='/primary-image.img'),
+                            MalformedImageHandler(number_of_failures=1, url='/primary-image.img')
+                            # TODO: this test fails too, although httpclient.cc sets
+                            # CURLOPT_LOW_SPEED_TIME and CURLOPT_LOW_SPEED_TIME
+                            #SlowRetrievalHandler(url='/primary-image.img')
```

Comment:
Do you know why this test fails? I didn't see a ticket related to this, but it's worth investigating and understanding.

Reply:
No, I don't. I didn't create ticket for this issue because I am not sure if this is an issue or not. I remember testing this case manually and IIRC Aktualizr did detect `slow retrieval attack`. I think, it might to do with specifics of using sockets on `localhost`. If it's so then it's not possible to test the given case in scope of the given test suite at all. Perhaps, it makes sense to create a ticket in order to double check it manually and make recommendations to the QA and/or E2E QA team to come with a test verifying Aktualizr against `slow retrieval attack`.

Reply:
> Perhaps, it makes sense to create a ticket in order to double check it manually and make recommendations to the QA and/or E2E QA team to come with a test verifying Aktualizr against `slow retrieval attack`.

Yeah, it's worth a ticket to follow up on this. It would be great if we could unit test it ourselves, but if it isn't possible, then a higher level is an option as well. There's a tech debt task for writing a test to catch oversized metadata, and this seems somewhat similar, albeit more challenging.

Reply:
https://saeljira.it.here.com/browse/OTA-3737